Implement bundle

apfelbox opened the pull request:
The initial implementation.

See the README for details.

Commit:
Implement bundle

keichinger commented on README.md:
```diff
@@ -0,0 +1,65 @@
+            - abc.*
+        backend:
+            - *.error
+    
```
```suggestion

```

keichinger commented on README.md:
````diff
@@ -0,0 +1,65 @@
+```
+
+The `extract` key defines which messages from which domain need to be extracted and dumped for usage in JavaScript.
+It is a nested array, where the key is the domain and the values are the message keys. You can use `*` as a placeholder.
````
Do we need to explain whether it's possible to dump an entire domain?

apfelbox replied:
Don‘t think so. 

Should be pretty straight forward to understand that you can just pass „*“

keichinger commented on src/Controller/DumpTranslationsCommand.php:
```diff
@@ -0,0 +1,42 @@
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+class DumpTranslationsCommand extends AbstractController
```
Can we rename it to `Controller`?

apfelbox replied:
Oopsie

keichinger commented on src/Controller/DumpTranslationsCommand.php:
```diff
@@ -0,0 +1,42 @@
+        $isDebug = $parameters->get("kernel.debug");
+        $catalogue = $extractor->fetchCatalogue($locale, !$isDebug);
+
+        $response = new Response(
```
Symfony has built-in support for JSONP with automatic output escaping (added comment).

```php
$response = new JsonResponse(catalogue->getCatalogueJson());
$response->setCallback("window.TranslatorInit.init");
```

E.g.
![image](https://user-images.githubusercontent.com/439899/61885376-06c4d580-aefe-11e9-99a2-bf3e399bedeb.png)

becomes
![image](https://user-images.githubusercontent.com/439899/61885401-13492e00-aefe-11e9-857f-82171cba0bf2.png)

apfelbox replied:
Will change

apfelbox replied:
We still need the wrapping for JSON, so that might look a bit weird.

apfelbox replied:
That is no output escaping, but a prevention mechanism for magic byte insertion, btw.

keichinger replied:
> That is no output escaping, but a prevention mechanism for magic byte insertion, btw.

True, true. I was referring to the added JavaScript comment, which is, like you said, no escaping. :)

keichinger commented on src/Extractor/TranslationsExtractor.php:
```diff
@@ -0,0 +1,134 @@
+
+        return $useCache
+            ? $this->cache->get(self::CACHE_PREFIX . ".{$locale}", $fetchCallback)
+            : $fetchCallback();
```
Do we need to catch the exception to re-throw a custom exception?

apfelbox replied:
We could validate the domain, but otherwise the exception can’t occur.

apfelbox replied:
I added the validation on the route - this is the only way to this method.

Commit:
Remove whitespace.

Commit:
Smaller issues fixed

+ Fixed controller name
+ Added requirements for locale
+ use native Symfony JsonResponse

Commit:
Add support for fallback catalogues

Commit:
Remove unnecessary flag

apfelbox commented:
I also added support for fallback locales, so `de-DE` will first look for every `de` translations, then overwrite with the specialized `de-DE`.
